File: scripts/repair_assets.js
const fs = require('fs');
const path = require('path');

const filePath = path.join(__dirname, '../public/data/assets.json');
let content = fs.readFileSync(filePath, 'utf8');

console.log("Original file size:", content.length);

// 1. Fix double commas
// Replace "}, , {" or "},, {" with "}, {"
content = content.replace(/},\s*,+\s*{/g, '}, {');

let data;
try {
    data = JSON.parse(content);
} catch (e) {
    console.log("Parse error, attempting more cleanup:", e.message);
    // Remove potential trailing commas before closing brackets
    content = content.replace(/,(\s*[\]}])/g, '$1');
    try {
        data = JSON.parse(content);
    } catch (e2) {
        console.error("Critical parse error:", e2.message);
        process.exit(1);
    }
}

const allowedCategories = ["GX", "未来都市", "モビリティ", "テクノロジー", "宇宙", "水中", "資源・バイオ", "エコ・ライフ"];

const categoryMap = {
    "スマートシティ": "未来都市",
};

data = data.map(item => {
    // 2. Image Path Sync
    const id = item.id;
    // Encode space and brackets: space -> %20, ( -> %28, ) -> %29
    const encodedId = id.replace(/ /g, '%20').replace(/\(/g, '%28').replace(/\)/g, '%29');

    // Folder naming convention based on prefix
    let prefix = id.split('-')[0].split(' ')[0].toLowerCase();

    // Especial case for id like "g-10 (2)" -> prefix is "g"
    item.url = `/assets/images/${prefix}/${encodedId}.jpg`;

    // 3. Category Normalization
    if (categoryMap[item.category]) {
        item.category = categoryMap[item.category];
    }

    if (!allowedCategories.includes(item.category)) {
        if (item.category.includes("都市")) item.category = "未来都市";
        else if (item.category.includes("エネ") || item.category.includes("GX")) item.category = "GX";
        else if (item.category.includes("水中") || item.category.includes("海")) item.category = "水中";
        else if (item.category.includes("宇宙")) item.category = "宇宙";
        else if (item.category.includes("車") || item.category.includes("交通") || item.category.includes("モビリティ")) item.category = "モビリティ";
        else if (item.category.includes("バイオ") || item.category.includes("資源")) item.category = "資源・バイオ";
        else if (item.category.includes("ライフ") || item.category.includes("エコ")) item.category = "エコ・ライフ";
        else item.category = "テクノロジー";
    }

    return item;
});

const output = JSON.stringify(data, null, 2);
fs.writeFileSync(filePath, output, 'utf8');
console.log("Successfully repaired assets.json");
console.log("Total items:", data.length);

File: scripts/repair_assets.py
import json
import re
import os

file_path = os.path.join(os.getcwd(), 'public/data/assets.json')

with open(file_path, 'r', encoding='utf-8') as f:
    content = f.read()

print(f"Original file size: {len(content)}")

# 1. Fix double commas
# Replace "}, , {" or "},, {" with "}, {"
content = re.sub(r'},\s*,+\s*{', '}, {', content)

try:
    data = json.loads(content)
except json.JSONDecodeError as e:
    print(f"Parse error, attempting more cleanup: {e}")
    # Remove potential trailing commas before closing brackets
    content = re.sub(r',(\s*[\]}])', r'\1', content)
    try:
        data = json.loads(content)
    except json.JSONDecodeError as e2:
        print(f"Critical parse error: {e2}")
        exit(1)

allowed_categories = ["GX", "未来都市", "モビリティ", "テクノロジー", "宇宙", "水中", "資源・バイオ", "エコ・ライフ"]

category_map = {
    "スマートシティ": "未来都市",
}

for item in data:
    # 2. Image Path Sync
    item_id = item.get('id', '')
    # Encode space and brackets: space -> %20, ( -> %28, ) -> %29
    encoded_id = item_id.replace(' ', '%20').replace('(', '%28').replace(')', '%29')
    
    # Folder naming convention based on prefix (everything before the first dash)
    # Example: g-1 -> g, mid-1 -> mid
    parts = item_id.split('-')
    if len(parts) > 1:
        prefix = parts[0].strip().split(' ')[0].lower()
    else:
        prefix = item_id.lower()
    
    item['url'] = f"/assets/images/{prefix}/{encoded_id}.jpg"

    # 3. Category Normalization
    category = item.get('category', 'テクノロジー')
    if category in category_map:
        category = category_map[category]
    
    if category not in allowed_categories:
        if "都市" in category: category = "未来都市"
        elif "エネ" in category or "GX" in category: category = "GX"
        elif "水中" in category or "海" in category: category = "水中"
        elif "宇宙" in category: category = "宇宙"
        elif "車" in category or "交通" in category or "モビリティ" in category: category = "モビリティ"
        elif "バイオ" in category or "資源" in category: category = "資源・バイオ"
        elif "ライフ" in category or "エコ" in category: category = "エコ・ライフ"
        else: category = "テクノロジー"
    
    item['category'] = category

with open(file_path, 'w', encoding='utf-8') as f:
    json.dump(data, f, ensure_ascii=False, indent=2)

print("Successfully repaired assets.json")
print(f"Total items: {len(data)}")
